examples/shamir-secret-share.masm

```diff
@@ -0,0 +1,150 @@
+#! - Instead of computing x^i for each iteration, store x^i in a variable and multiply by x to get next exponent.
+
+
+# create secret commitment (hash of secret + rand)
+proc.gen_secret_com
+    # input: [secret, rand]
+    padw
+    push.0.0      #[0, 0, 0, 0, 0, 0, secret, rand]
+    rphash        #[com_3, com_2, com_1, com_0]
+end
+
+# create k-1 random polynomial coefficients
+proc.gen_poly_coeffs
+    # input [k, rand]
+
+    # generate initial 4 random coefficients from rand
+    # coef_0 = secret
+    swap    #[rand, k]
+    padw    #[0, 0, 0, 0, rand, k]
+    padw    #[0, 0, 0, 0, 0, 0, 0, 0, rand, k]
+    drop    #[0, 0, 0, 0, 0, 0, 0, rand, k]
+    rphash  #[coef_4, coef_3, coef_2, coef_1, k]
+
+    # introduce counter and evaluate if we have enough coefficients
+    # counter is initialized to 5 as we must account for the secret which acts as the intercept
+    push.5  #[counter, coef_4, coef_3, coef_2, coef_1, k]
+    movup.5 #[k, counter, coef_4, coef_3, coef_2, coef_1]
+    dup.1
+    dup.1
+    lt
+
+    # continue creating more coefficients until we have enough (created in 4's)
+    while.true     #[k, counter, coef_4, coef_3, coef_2, coef_1]
+        # generate more coefficients by padding and hashing previous coefficients
+        movdn.5
+        movdn.5
+        dupw       #[coef_4, coef_3, coef_2, coef_1, coef_4, coef_3, coef_2, coef_1, k, counter, ...]
+        padw       #[0, 0, 0, 0, coef_4, coef_3, coef_2, coef_1, coef_4, coef_3, coef_2, coef_1, k, counter]
+        rphash     #[coef_8, coef_7, coef_6, coef_5, coef_4, coef_3, coef_2, coef_1, k, counter, ...]
+
+        # evaluate if we have enough coefficients
+        movup.9
+        add.4
+        movup.9
+        dup.1
+        dup.1
+        lt
+    end         #[k, counter, coef_k-1, ...]
+    
+    # evaluate if we have too many coefficients
+    dup          #[k, k, counter, coef_k-1, ...]
+    movdn.2      #[k, counter, k, coef_k-1, ...]
+    sub          #[diff, k, coef_k-1, ...]
+    dup          #[diff, diff, k, coef_k-1, ...]
+    push.0       #[0, diff, diff, k, coef_k-1, ...]
+    gt           #[1, diff, k, coef_k-1] 
+
+    # drop unrequired coefficients such that we have k in total
+    while.true #[diff, k, coef_k-1]
+        movup.2
+        drop
+        sub.1
+        dup
+        push.0
+        gt
+    end
+
+    drop
+end
+
+# evaluate polynomial
+proc.eval_poly
+    #[x, k, coef_k-1,...]
+
+    # introduce sum variable to calculate y
+    swap    #[k, x, coef_k-1, ..., secret]
+    push.0  #[sum, k, x, coef_k-1, ..., secret]
+    push.1  #[1, sum, k, x, coef_k-1, ..., secret]
+
+    # evaluate polynomial by iterating through terms and summing
+    while.true 
+        dup.2 #[x, sum, k, x, coef_k-1,..., secret]
+        dup.2 #[k, x, sum, k, x, coef_k-1,..., secret]
+        sub.1 #[k-1, x, sum, k, x, coef_k-1,..., secret]
+        exp   #[x^k-1, sum, k, x, coef_k-1,..., secret]
```

Comment:
nit: `x^(k-1)` here and below. It might be clearer to use parens for `coef_(k-1)` as well